### ISSUE TITLE
Databricks Workspace - add support for configuring Customer Managed Keys for DBFS

### ISSUE THREAD

**Issue by alexott** [hashicorp/terraform-provider-azurerm#14420]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Right now, terraform provider only supports enabling of [double encryption for DBFS Root](https://docs.microsoft.com/en-us/azure/databricks/security/keys/double-encryption), but it doesn't support specification of [customer managed keys for DBFS Root](https://docs.microsoft.com/en-us/azure/databricks/security/keys/customer-managed-keys-dbfs/) that is a separate feature of Databricks workspace

### New or Affected Resource(s)

* `azurerm_databricks_workspace`

### Potential Terraform Configuration

```hcl
resource "azurerm_databricks_workspace_customer_managed_key" "example" {
  depends_on = [azurerm_key_vault_access_policy.databricks]

  workspace_id     = azurerm_databricks_workspace.example.id
  key_vault_key_id = azurerm_key_vault_key.example.id
}

resource "azurerm_databricks_workspace" "example" {
....
  dbfs_cmk_key_vault_key_id = azurerm_key_vault_key.example.id
}
```

**Comment by alexott** [comment]
I was thinking about the same approach as for CMK for Managed Services

**Comment by atombravo** [comment]
https://github.com/hashicorp/terraform-provider-azurerm/blob/main/examples/databricks/customer-managed-key/dbfs/main.tf There is support (recently added) for customer managed keys for the DBFS (though the AzureRM provide docs are rather unclear about which "customer managed key" is which).  There is a separate resource, https://registry.terraform.io/providers/hashicorp/azurerm/2.84.0/docs/resources/databricks_workspace_customer_managed_key
as @tombuildsstuff  suggested.

It currently won't work with infrastructure encryption aka double encryption, because the provider has a check that prevents that (no idea why).. I just filed a related feature request this morning asking to remove that check, since there's no reason to prevent them both from being enabled (you can do it from the AZ command line) : https://github.com/hashicorp/terraform-provider-azurerm/issues/14424

I

**Comment by alexott** [comment]
From docs it’s not clear that this resource is used for DBFS, because docs also refers for managed services, aka notebooks… 
Imho it would be easier to have it similar to managed services. Let’s try to remove confusion, because it’s really hard to understand

**Comment by atombravo** [comment]
The documentation is *really* confusing, not helped by the fact that the Microsoft documentation uses different terms in their documentation... it's a jumble.  My guess is that the naming of this was chosen by the team writing the provider when there was only DBFS encryption available (and it would have been customer managed keys or Microsoft managed keys).. and then Microsoft added more optional encryption which used the same terms that the provider had been using.

So.. four layers of possible encryption:
1. DBFS databrick file system key (called the customer managed key in TF docs)
2. Managed services key (added by your request: https://github.com/hashicorp/terraform-provider-azurerm/issues/12302)
3. Infrastructure encryption key (for "double encryption" of DBFS) - Microsoft manages keys, we don't touch them (and currently you can't enable it with TF if you enable the DBFS)
4. OTW On the wire encryption.. which is magic that happens inside the OS/spark config -- not in Terraform (yet?)

**Comment by alexott** [comment]
Yes, you're correct. Really, it would be nice to unify the things, maybe get rid of the separate resource, and instead use blocks inside workspace, or instead re-use the separate resource and use it for managed services as well by allowing to specify for which use-case it will be applied (similarly to what is done for [AWS](https://registry.terraform.io/providers/databrickslabs/databricks/latest/docs/resources/mws_customer_managed_keys)) - new APIs are designed to be extensible, and I can foresee that new use-cases for CMK will be coming.

But in the short term, it would be nice to fix documentation - I'll try to submit PR.

**Comment by nathanknox** [comment]
I know it's been a while, just bumping here to see if there's an update on this issue. Thanks in advance!

**Comment by alexott** [comment]
@favoretti - this could be closed btw

**Comment by alexott** [comment]
It's supported for a long time already - since 2.67.0. Update of existing workspace is supported since version 3.41.0. Upgrade needs to be done in few steps:

1. set `customer_managed_key_enabled` to `true` and apply
2. add AKV access policy for the workspace's storage identity
3. add `azurerm_databricks_workspace_customer_managed_key` resource with the key and apply

Steps 2 & 3 could be joined together, but will require explicit dependency between access policy and `azurerm_databricks_workspace_customer_managed_key`
